stackrox/stackrox
# ROX-18330: Invalidate tokens after auth provider changes

## Description

In case of changes to the auth provider _where the ID of the provider will not change_, i.e. within the declarative configuration approach, Central will currently not _force_ people to relogin.

This means two things:
- changes made with declarative config may not immediately apply and be visible without using ensuring you use a _fresh_ session.
- previously authenticated and authorized clients will keep their existing roles for the remaining TTL of their token.

Instead, we should force a relogin in case the configuration changed. This will ensure changes will be propagated accordingly to all existing users by forcing them to establish a new session, thus ensuring they
have the correct roles assigned.

To establish this, the following has been done:
- Introduce a new field `LastUpdated` which indicates the last updated made to an auth provider.
- Ensured that on both `Create` and `Update` the last updated field is updated to the current time.
- Added a validation within the provider to compare the claim `issued_at` with the `LastUpdated` field.

Few things to note:
- In case the field `LastUpdated` isn't set explicitly, it will assume the zero time. Thus, no token will be seen as invalid up until the auth provider is updated.
- Currently the scope of setting the `LastUpdated` field is limited to changes for the auth provider. In case groups are changed, no re-login will be enforced (albeit, this could be done quite easily).
- No migration is required for this change. The code can handle the default value of the field (guarded by unit test) and the schema of the table doesn't change (the `LastUpdated` field isn't searchable, thus no separate column).

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

- see added unit tests.

Manual tests:
1. Configure declarative configuration providing an auth provider.

2. Verify login works successfully.

3. When querying the `/v1/authProviders` API, the auth provider should have a `lastUpdated` field set.

3. Update the auth provider by changing either some claim mapping, client ID/secret or any other field of your choice (_except_ the name).

4. Within the `/v1/authProviders` API, verify that the `lastUpdated` value is more recent than before.

5. Re-open the browser tab where the user logged in previously. Within the developer tool, you should see some requests are failing with 403 or 401:
![Screenshot 2023-08-22 at 02 19 42](https://github.com/stackrox/stackrox/assets/89904305/958bd3a6-3a3d-432f-a89e-aac366f3e626)

6. Verify the `iat` field of the rox token (you can retrieve it from the request headers in the developer tool:
![Screenshot 2023-08-22 at 02 23 18](https://github.com/stackrox/stackrox/assets/89904305/82f8eb4d-6d79-4d92-b0c1-ec1f31553b5a)

7. Verify that the UI _either_ will prompt to re-login in case no refresh capability is set for the auth provider _or_ that the `tokenrefresh` API is called
![Screenshot 2023-08-22 at 02 20 54](https://github.com/stackrox/stackrox/assets/89904305/f6f4b215-6339-46d7-ab61-587bba2ee198)

8. Inspect the token used afterwards and notice that the `iat` (issued at) field is now updated to the _current_ time:
![Screenshot 2023-08-22 at 02 24 03](https://github.com/stackrox/stackrox/assets/89904305/6202449d-2609-4d3c-94a3-f84e18c17f22)
